### ISSUE TITLE
Things to change

### ISSUE THREAD

**Issue by FinnStutzenstein** [OpenSlides/openslides3-autoupdate-service#16]
- start with an empty cache (=flushall): `Can not initialize data: get startdata from redis: invalid values in max change id responce. Got 0, expected 2`
- This error happens after some time. Mostly if a client reconnects, but this just popped up all the time, so I'm not sure, what are the root causes.
```
Error: Can not receive new data: receive missing data from 135235678000 to 135235678001: get data: hmget all request: redigo: StringMap expects even number of values result
topic id differs from change id. Topic: 135235678001, changeid 0
panic: topic id differs from change id. Topic: 135235678001, changeid 0
```
This happens, if the cache gets cleared:
```
Error: Can not receive new data: get autoupdate data: read autoupdate from redis: EOF
Error: Can not receive new data: receive missing data from 135238608000 to 135239173000: get data: hmget all request: redigo: StringMap expects even number of values result
```
-> The service must be secured agains random cache flushes, invalid data in redis and loosing connections.

- Chromium kills the connection after some time (in both cases, I didn't get any headers or so):
  * If there was never a message (since the client is up to date) after 240s the connection gets closed:
```
net::ERR_EMPTY_RESPONSE

status: 0
statusText: "Unknown Error"
url: "/system/autoupdate"
ok: false
name: "HttpErrorResponse"
message: "Http failure response for /system/autoupdate: 0 Unknown Error"
```
  * If there was a message, after 120s the connection is closed, if no new message arrived:
```
net::ERR_INCOMPLETE_CHUNKED_ENCODING 200 (OK)

status: 0
statusText: "Unknown Error"
url: "http://localhost:4200/system/autoupdate?change_id=135252742000"
ok: false
name: "HttpErrorResponse"
message: "Http failure response for http://localhost:4200/system/autoupdate?change_id=135252742000: 0 Unknown Error"
```
I'm not sure about other browsers, but I can see the reason: Since this is still a "normal" http request, it is closed, if there is no activity. See: https://tools.ietf.org/id/draft-loreto-http-bidirectional-07.html
-> We have two possibilities:
1) Send a layer-7 keep alive. E.g. a message with the key `keep-alive`, which gets discarded in the client code.
2) The client always must reopen the connection.

Solution 1 is nearer to the old behavior. Solution 2 would mean a complete rewrite of the communication stack, since it is builded upon the idea of having steady connections and an abort is something unusual. It would be nice, if you implement keep alive messages.

TL;DR:
- The change_id behavior is broken. I couldn't test restricters, because this was too annoying
- The service must be robust against enviromental failures
- We need keep-alive messages, so the connections don't get killed by browsers.

**Comment by ostcar** [comment]
>     * start with an empty cache (=flushall): `Can not initialize data: get startdata from redis: invalid values in max change id responce. Got 0, expected 2`
See #18 

>     * This error happens after some time. Mostly if a client reconnects, but this just popped up all the time, so I'm not sure, what are the root causes.
> 
> 
> ```
> Error: Can not receive new data: receive missing data from 135235678000 to 135235678001: get data: hmget all request: redigo: StringMap expects even number of values result
> topic id differs from change id. Topic: 135235678001, changeid 0
> panic: topic id differs from change id. Topic: 135235678001, changeid 0
> ```
This should be fixed with #17 

> This happens, if the cache gets cleared:
> 
> ```
> Error: Can not receive new data: get autoupdate data: read autoupdate from redis: EOF
> Error: Can not receive new data: receive missing data from 135238608000 to 135239173000: get data: hmget all request: redigo: StringMap expects even number of values result
> ```
> -> The service must be secured agains random cache flushes, invalid data in redis and loosing connections.
Partly fixed with #17, the rest is fixed local and will be pushed soon. I don't know how to test invalid data

>     * Chromium kills the connection after some time (in both cases, I didn't get any headers or so):
Moved to #19

>     * The change_id behavior is broken. I couldn't test restricters, because this was too annoying
Please try again

>     * The service must be robust against enviromental failures
What do you mean? How can I test it?

**Comment by ostcar** [comment]
See https://github.com/golang/go/issues/31643

It seems, that the go-server sends [http2 ping-frames](https://http2.github.io/http2-spec/#PING) per default. Could you test you issue with curl and the `--http2` flag. Maybe it is just an issue of http 1.1?

**Comment by FinnStutzenstein** [comment]
> Maybe it is just an issue of http 1.1?

Even if it in http/1.1 issue, we need to make this work during development. If we can test this in production with http/2, we can disable the KA if we want.

**Comment by ostcar** [comment]
I close this meta issue. We can discuss the problems in the other issues.
